File: blanc/src/algorithm/leetcode/p201to250/LeetCode208ImplementTrie.java
package algorithm.leetcode.p201to250;

@SuppressWarnings("unused")
public class LeetCode208ImplementTrie {

    private interface Trie {

        /**
         * Inserts a word into the trie.
         */
        void insert(String word);

        /**
         * Returns if the word is in the trie.
         */
        boolean search(String word);

        /**
         * Returns if there is any word in the trie that starts with the given prefix.
         */
        boolean startsWith(String prefix);
    }

    private static class Impl implements Trie {

        private final Node root = new Node();

        @Override
        public void insert(String word) {
            Node node = root;
            for (char c : word.toCharArray()) {
                int index = c - 'a';
                if (node.next[index] == null) {
                    node.next[index] = new Node();
                }
                node = node.next[index];
            }
            node.isWord = true;
        }

        @Override
        public boolean search(String word) {
            Node node = root;
            for (char c : word.toCharArray()) {
                int index = c - 'a';
                if (node.next[index] == null) {
                    return false;
                }
                node = node.next[index];
            }
            return node.isWord;
        }

        @Override
        public boolean startsWith(String prefix) {
            Node node = root;
            for (char c : prefix.toCharArray()) {
                int index = c - 'a';
                if (node.next[index] == null) {
                    return false;
                }
                node = node.next[index];
            }
            return true;
        }

        private static class Node {
            Node[] next = new Node[26];
            boolean isWord = false;
        }
    }
}

File: src/leetcode/p201to250/LeetCode208ImplementTrie.kt
package leetcode.p201to250

class LeetCode208ImplementTrie {

    interface Trie {
        /** Inserts a word into the trie. */
        fun insert(word: String)

        /** Returns if the word is in the trie. */
        fun search(word: String): Boolean

        /** Returns if there is any word in the trie that starts with the given prefix. */
        fun startsWith(prefix: String): Boolean
    }

    class Solution : Trie {

        class TrieNode (private var value: Char) {
            var next = arrayOfNulls<TrieNode>(26)
            var isWord: Boolean = false
        }

        /** Initialize your data structure here. */
        private val root = TrieNode('a')
        private val STANDARD = 'a'.toInt()

        override fun insert(word: String) {
            val chars = word.toCharArray()
            var node = root
            for (char in chars) {
                val num = char.toInt() - STANDARD
                node = node.next[num]?: TrieNode(char).apply { node.next[num] = this}
            }
            node.isWord = true
        }

        override fun search(word: String): Boolean {
            val chars = word.toCharArray()
            var node = root
            for (char in chars) {
                val num = char.toInt() - STANDARD
                val next = node.next[num]
                if (next == null) {
                    return false
                } else {
                    node = next
                }
            }
            return node.isWord
        }

        override fun startsWith(prefix: String): Boolean {
            val chars = prefix.toCharArray()
            var node = root
            for (char in chars) {
                val num = char.toInt() - STANDARD
                val next = node.next[num]
                if (next == null) {
                    return false
                } else {
                    node = next
                }
            }
            return true
        }
    }
}